src/Kiota.Builder/Export/PublicAPIExportService.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Kiota.Builder.CodeDOM;
+using Kiota.Builder.Configuration;
+using Kiota.Builder.Writers;
+using Kiota.Builder.Writers.CSharp;
+using Kiota.Builder.Writers.Go;
+using Kiota.Builder.Writers.Java;
+using Kiota.Builder.Writers.Php;
+using Kiota.Builder.Writers.Python;
+using Kiota.Builder.Writers.Ruby;
+using Kiota.Builder.Writers.Swift;
+using Kiota.Builder.Writers.TypeScript;
+
+namespace Kiota.Builder.Export;
+
+internal class PublicApiExportService
+{
+    internal PublicApiExportService(GenerationConfiguration generationConfiguration)
+    {
+        ArgumentNullException.ThrowIfNull(generationConfiguration);
+        _outputDirectoryPath = generationConfiguration.OutputPath;
+        _conventionService = GetLanguageConventionServiceFromConfiguration(generationConfiguration);
+    }
+    private readonly string _outputDirectoryPath;
+    private readonly ILanguageConventionService _conventionService;
+    private const string DomExportFileName = "kiota-dom-export.txt";
+    private const string InheritsSymbol = "-->";
+    private const string ImplementsSymbol = "~~>";
+    private const string OptionalSymbol = "?";
+    private const string ParentElementAndChildSeparator = "::";
+
+    internal async Task SerializeDomAsync(CodeNamespace rootNamespace, CancellationToken cancellationToken = default)
+    {
+        var filePath = Path.Combine(_outputDirectoryPath, DomExportFileName);
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task
+        await using var fileStream = File.Create(filePath);
```

Comment:
Direct file access can be removed from this service and a stream provided instead. That will allow easier testing with memory streams. It will also enable providing a config for the output path.

Reply:
Moved this out in latest update.